rust/xaynet-server/src/state_machine/phases/update.rs

```diff
@@ -186,148 +186,370 @@ where
+            .with_update_count_min(1)
+            .with_update_count_max(1)
+            .with_update_time_min(1)
+            .with_update_time_max(5)
+            .build();
+
+        let (event_publisher, event_subscriber) = events_from_sum_phase(&state);
+        let events_before_update = EventSnapshot::from(&event_subscriber);
+        let state_before_update = state.clone();
+
+        let (shared, request_tx) = init_shared(state, store, event_publisher);
+        let state_machine = StateMachine::from(PhaseState::<Update, _>::new(shared));
+        assert!(state_machine.is_update());
+
+        send_update_messages(1, request_tx.clone());
+        let state_machine = state_machine.next().await.unwrap();
+
+        let state_after_update = state_machine.shared_state_as_ref().clone();
+        let events_after_update = EventSnapshot::from(&event_subscriber);
+        assert_after_phase_failure(
+            &state_before_update,
+            &events_before_update,
+            &state_after_update,
+            &events_after_update,
+        );
+
+        assert!(state_machine.is_failure());
+        assert!(matches!(
+            state_machine.into_failure_phase_state().private.error,
+            PhaseError::Update(UpdateError::NoSeedDict)
+        ))
+    }
+
+    #[tokio::test]
+    async fn test_aggregation_error() {
+        // No Storage errors
+        //
+        // What should happen:
+        // 1. broadcast Update phase
+        // 2. reject 3 update messages (validation of the models fail due to an invalid length)
+        // 3. accept 3 update messages
+        // 4. fetch seed dict
+        // 5. broadcast seed dict
+        // 6. move into sum2 phase
+        //
+        // What should not happen:
+        // - the shared state has been changed
+        // - the global model has been invalidated
+        // - the sum dict has been invalidated
+        enable_logging();
+
+        let mut cs = MockCoordinatorStore::new();
+        cs.expect_add_local_seed_dict()
+            .times(3)
+            .returning(move |_, _| Ok(LocalSeedDictAdd(Ok(()))));
+        cs.expect_seed_dict()
+            .return_once(move || Ok(Some(SeedDict::new())));
+        let store = Store::new(cs, MockModelStore::new());
+        let state = CoordinatorStateBuilder::new()
+            .with_round_id(1)
+            .with_update_count_min(3)
+            .with_update_count_max(3)
+            .with_update_time_min(1)
+            .build();
+
+        let (event_publisher, event_subscriber) = events_from_sum_phase(&state);
+        let events_before_update = EventSnapshot::from(&event_subscriber);
+        let state_before_update = state.clone();
+
+        let (shared, request_tx) = init_shared(state, store, event_publisher);
+        let state_machine = StateMachine::from(PhaseState::<Update, _>::new(shared));
+        assert!(state_machine.is_update());
+
+        send_update_messages_with_model(3, request_tx.clone(), create_mask(2, 1));
```

Comment:
even though calling this function with a mask instead of a masked model works because both are of `MaskObject` type, this is just another reason to finally have distinct wrapper types for masks and masked models. we should address this in a future refactoring.

Reply:
I agree👍 Do we have a ticket/story for that?

Reply:
i don't think that we already have a ticket, can you create one please?

Reply:
I can do that 👍